hknweb/candidate/views.py

```diff
@@ -21,50 +48,87 @@ class CandRequestView(FormView, generic.ListView):
     context_object_name = 'challenge_list'
 
     def form_valid(self, form):
-        # This method is called when valid form data has been POSTed.
-        # It should return an HttpResponse.
         form.instance.requester = self.request.user
         form.save()
-        self.send_email(form)
+        self.send_request_email(form)
+        messages.success(self.request, 'Your request was submitted to the officer!')
         return super().form_valid(form)
 
-    def send_email(self, form):
+    def send_request_email(self, form):
         subject = 'Confirm Officer Challenge'
         officer_email = form.instance.officer.email
         text_content = 'Confirm officer challenge'
 
-        candidate_name = form.instance.requester.get_full_name()
-        candidate_username = form.instance.requester.username
-        # host = self.request.get_host()
-        link = self.request.build_absolute_uri(
+        confirm_link = self.request.build_absolute_uri(
                 reverse("candidate:challengeconfirm", kwargs={ 'pk' : form.instance.id }))
         html_content = render_to_string(
-            'candidate/email.html',
+            'candidate/request_email.html',
             {
-                'pk': form.instance.id,
-                'candidate_name' : candidate_name,
-                'candidate_username' : candidate_username,
-                'link' : link,
+                'candidate_name' : form.instance.requester.get_full_name(),
+                # TODO: for some usernames such as catherine.hu, this becomes a link. Why??
+                'candidate_username' : form.instance.requester.username,
+                'confirm_link' : confirm_link,
+                'img_link' : get_rand_photo(),
             }
         )
         msg = EmailMultiAlternatives(subject, text_content,
                 'no-reply@hkn.eecs.berkeley.edu', [officer_email])
         msg.attach_alternative(html_content, "text/html")
         msg.send()
 
-    def get_context_data(self, **kwargs):
-        context = super(CandRequestView, self).get_context_data(**kwargs)
-        return context
-
     def get_queryset(self):
-        result = OffChallenge.objects
+        result = OffChallenge.objects \
+                .filter(requester__exact=self.request.user) \
+                .order_by('-request_date')
+        return result
+
+
+# List of past challenge requests for officer
+# Non-officers can still visit this page but it will not have any entries
+@method_decorator(login_required(login_url='/accounts/login/'), name='dispatch')
```

Comment:
Is there a reason this is a @method_decorator, and not actually a decorator on a method?

Reply:
Decorators work on functions. To decorate an object method, you need a method decorator